p2p/memp2p/messenger.go

```diff
@@ -264,100 +281,101 @@ func (messenger *Messenger) BroadcastOnChannelBlocking(channel string, topic str
 // parametricBroadcast() is done as a go-routine, which means this function is,
 // in fact, non-blocking, but it is identical with BroadcastOnChannelBlocking()
 // in all other regards.
-func (messenger *Messenger) BroadcastOnChannel(channel string, topic string, buff []byte) {
-	err := messenger.parametricBroadcast(topic, buff, false)
+func (messenger *Messenger) BroadcastOnChannel(_ string, topic string, buff []byte) {
+	err := messenger.synchronousBroadcast(topic, buff)
 	log.LogIfError(err)
 }
 
 // Broadcast asynchronously sends the message to all peers in the network. It
 // calls parametricBroadcast() with async=true, which means that peers will
 // have their ReceiveMessage() function independently called as go-routines.
 func (messenger *Messenger) Broadcast(topic string, buff []byte) {
-	err := messenger.parametricBroadcast(topic, buff, true)
+	err := messenger.synchronousBroadcast(topic, buff)
 	log.LogIfError(err)
 }
 
-// parametricBroadcast sends a message to all peers in the network, with the
-// possibility to choose from asynchronous or synchronous sending.
-func (messenger *Messenger) parametricBroadcast(topic string, data []byte, async bool) error {
+// synchronousBroadcast sends a message to all peers in the network in a synchronous way
+func (messenger *Messenger) synchronousBroadcast(topic string, data []byte) error {
 	if !messenger.IsConnectedToNetwork() {
 		return ErrNotConnectedToNetwork
 	}
 
-	message, err := NewMessage(topic, data, messenger.ID())
-	if err != nil {
-		return err
+	seqNo := atomic.AddUint64(&messenger.seqNo, 1)
+	message := newMessage(topic, data, messenger.ID(), seqNo)
+
+	peers := messenger.network.Peers()
+	for _, peer := range peers {
+		peer.receiveMessage(message)
 	}
 
-	for _, peer := range messenger.Network.Peers() {
-		if async {
-			go func(receivingPeer *Messenger) {
-				err := receivingPeer.ReceiveMessage(topic, message, true)
-				log.LogIfError(err)
-			}(peer)
-		} else {
-			err = peer.ReceiveMessage(topic, message, true)
+	return nil
+}
+
+func (messenger *Messenger) processFromQueue() {
+	for {
+		message := <-messenger.processQueue
+		if check.IfNil(message) {
+			continue
 		}
+
+		identifier := string(message.Peer()) + string(message.SeqNo())
+		messenger.seenMessages.Sweep()
+		err := messenger.seenMessages.Add(identifier)
 		if err != nil {
-			break
+			log.Warn("reply back")
+			return
 		}
-	}
 
-	return err
+		topic := message.TopicIDs()[0]
+		if topic == "" {
+			continue
+		}
+
+		messenger.topicsMutex.Lock()
+		_, found := messenger.topics[topic]
+		if !found {
+			messenger.topicsMutex.Unlock()
+			continue
+		}
+
+		// numReceived gets incremented because the message arrived on a registered topic
+		atomic.AddUint64(&messenger.numReceived, 1)
+		validator, _ := messenger.topicValidators[topic]
+		if check.IfNil(validator) {
+			messenger.topicsMutex.Unlock()
+			continue
+		}
+		messenger.topicsMutex.Unlock()
```

Comment:
Move this before the `if` and only do it once.

Reply:
this is not possible as I need the same mutex for the second map that holds topicValidators. I want the check to be done atomically.